### PR TITLE
Static library linking support

### PR DESCRIPTION
**Pull request by partouf**
More of a brainstorm PR than a complete solution.
Also wanted to add dependency support, but not sure how to do it.

On the other hand, I'm wondering we shouldn't just filter out links from user arguments and put them in the right place, that way we wouldn't have to add more and more libraries just for this.

Ideas?

Todo
- [x] Add dependencies property (pointing to other static libs?)
   - This is for ordering of all -l statements that are linking to static libs, but they still have to be selected in CE's UI, so this doesn't auto-select dependent libraries.
- [x] Do + test ordering, eliminate duplicates?

### CODE REVIEW

File: test/options-handler.js [hunk 1]
```diff
@@ -155,4 +173,19 @@ describe('Options handler', () => {
         });
         optionsHandler.setCompilers([]);
     });
+    it('should get static libraries', () => {
```
Comment (mattgodbolt) [inline test/options-handler.js]: this is great! Thanks :)

One note that confused me to start with: I think of "static link" as meaning `-static` but I don't know if that's what's happening here? We're linking with the default (which will be dynamic linking I think), right?

File: test/options-handler.js [hunk 2]
```diff
@@ -91,13 +95,27 @@ describe('Options handler', () => {
...
+                "description": undefined,
+                "name": undefined,
+                "url": undefined,
+                "versions": {
+                    "std": {
+                        "libpath": [],
+                        "path": [],
+                        "version": "std",
+                        "liblink": [],
+                        "staticliblink": ["rt", "pthread", "c++fs"]
```
Comment (mattgodbolt) [inline test/options-handler.js]: :+1: love this

File: test/options-handler.js (hunk 1 above)
Comment (partouf) [reply test/options-handler.js]: This is more about linking to .a's into your binary, which doesn't really need supplying -static. But -static could be a good idea to play with as well.

File: lib/base-compiler.js [hunk 3]
```diff
@@ -210,6 +210,83 @@ class BaseCompiler {
...
+                    libCompareObj.dependencies.includes(libToInsertName)) {
+                    continue;
+                } else if (libToInsertObj &&
+                    libToInsertObj.staticliblink.includes(libToInsertName) &&
+                    libToInsertObj.staticliblink.includes(libCompareName)) {
+                    if (libToInsertObj.staticliblink.indexOf(libToInsertName) >
+                        libToInsertObj.staticliblink.indexOf(libCompareName)) {
+                        idxToInsert = idx + 1;
+                    } else {
+                        idxToInsert = idx;
```
Comment (partouf) [inline lib/base-compiler.js]: test this

File: lib/base-compiler.js [hunk 4]
```diff
@@ -210,6 +210,83 @@ class BaseCompiler {
...
+                    }
+                    break;
+                } else if (libCompareObj && libCompareObj.staticliblink.includes(libToInsertName) &&
+                    libCompareObj.staticliblink.includes(libCompareName)) {
```
Comment (partouf) [inline lib/base-compiler.js]: test this

**Comment by partouf** [comment]
Maybe Todo still is; somewhere along the way I lost information in the compilationOptions that is returned to the client. It doesn't contain -lstdc++fs after the filename?

**Pull request merged**